### ISSUE TITLE
Partial substitutions of Mul objects do not work for mixed-variable terms

### ISSUE THREAD

**Issue by tbabej** [symengine/symengine#1359]
Consider the following example, which works as expected:

```python
>>> import symengine
>>> x,y,z = symengine.symbols('x y z')
>>> eq = 2*x*y*y + 3*x*x + 4*y*y
>>> eq.subs(y*y, z)
4*z + 2*x*z + 3*x**2
```
If we try to substitute `x*y` instead of `y*y`, which is still contained in the first term, the substitution does not happen:

```python
>>> eq.subs(x*y, z)
2*x*y**2 + 3*x**2 + 4*y**2
```
This only happens for non-exact matches, exact matches do work:

```python
>>> (2*x*y + 3*y).subs(x*y, z)
3*y + 2*z
```

**Comment by isuruf** [comment]
Yes, `subs` is kind of dumb that way. What happens is it will check whether the expression is equal to `2*x*y**2`, then `x*y**2` then `x` then `y**2`. 
This should be fixed though. Original behaviour can stay in `xreplace`, but `subs` needs to be fixed.

**Comment by tbabej** [comment]
@isuruf Can you point me to the place in the code base where this logic is implemented? I can contribute a fix.

**Comment by isuruf** [comment]
Thanks. It's implemented in `XReplaceVisitor` at https://github.com/symengine/symengine/blob/master/symengine/subs.h#L74-L108

This method shouldn't be changed, but overriden in the `SubsVisitor` class which is here, https://github.com/symengine/symengine/blob/master/symengine/subs.h#L270

**Comment by eeshan9815** [comment]
@tbabej @isuruf Hey, I am new to symengine. If no one is working on it, I would like to start contributing by fixing this issue.

**Comment by ShikharJ** [comment]
Go for it @eeshan9815

**Comment by tbabej** [comment]
Indeed, we just settled for a python-based dirty quick workaround in our
code. It would be great to support this natively.

**Comment by eeshan9815** [comment]
@isuruf @ShikharJ @tbabej I think I found another issue with `subs`:
While this works as expected:
```
>>> (2*x*y + 3*y).subs(x*y, z)
3*y + 2*z
```
This does not:
```
>>> (2*x*y).subs(x*y, z)
2*x*y
```
instead of `2*z`
Can someone explain why, and how to fix this issue as well? I have had some progress with the original issue.
